# Assignment 2 - Data Visualization and Manipulation

This assignment will focus on two very important libaries in machinese learning and data science: Matplotlib and Pandas. You will learn several data manipulation and visualization techniques. However, there exists a lot more functions out there. Don't forget to get curious and explore what these powerful libraries are capable of!

*Don't forget that commenting your code is very important!*

### 1. Import packages

###### Importing packages and knowing what packages you need for a project is crucial. We will not be reminding you which packages you need for each question and for the assignment in general. Please import the packages at your own discretion. Although it is common practice to import all packages at once at the beginning, don't hesitate to revisit the next cell, and import more packages as you may need. 

*Note*: we presented some example packages during class (eg. numpy, matplotlib, etc.). However, we only focus on the end results, so you can use any libraries of you choice. 

In [2]:
# IMPORT PACKAGES
# =========== YOUR CODE HERE ======== 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### 2. Data Manipulation

For this part of the assignment, we will be working with ramen-ratings dataset from Kaggle. <br>
https://www.kaggle.com/residentmario/ramen-ratings 

###### Read the 'dataset/ramen-ratings.csv' dataset. Describe the dataset quickly. What do you see? 

In [4]:
# =========== YOUR CODE HERE ========
df=pd.read_csv('./dataset/ramen-ratings.csv')
df

Review #             Brand  \
0         2580         New Touch   
1         2579          Just Way   
2         2578            Nissin   
3         2577           Wei Lih   
4         2576    Ching's Secret   
5         2575     Samyang Foods   
6         2574           Acecook   
7         2573       Ikeda Shoku   
8         2572        Ripe'n'Dry   
9         2571              KOKA   
10        2570       Tao Kae Noi   
11        2569          Yamachan   
12        2568          Nongshim   
13        2567            Nissin   
14        2566            Nissin   
15        2565              KOKA   
16        2564              TRDP   
17        2563          Yamachan   
18        2562          Binh Tay   
19        2561             Paldo   
20        2560            Nissin   
21        2559  Jackpot Teriyaki   
22        2558              KOKA   
23        2557          Yamachan   
24        2556    Ching's Secret   
25        2555     Samyang Foods   
26        2554           Yum-Mie   
27        2553            Nissin   
28        2552           MyKuali   
29        2551              KOKA   
...        ...               ...   
2550        30             Myojo   
2551        29          Mee Jang   
2552        28            Nissin   
2553        27            Nissin   
2554        26            Nissin   
2555        25            Nissin   
2556        24            Nissin   
2557        23            Nissin   
2558        22          Nongshim   
2559        21          Nongshim   
2560        20          Nongshim   
2561        19          Nongshim   
2562        18            Ottogi   
2563        17         Quickchow   
2564        16           Samyang   
2565        15           Samyang   
2566        14           Samyang   
2567        13   Sapporo Ichiban   
2568        12   Sapporo Ichiban   
2569        11       Six Fortune   
2570        10             Smack   
2571         9             Sutah   
2572         8            Tung-I   
2573         7           Ve Wong   
2574         6             Vifon   
2575         5             Vifon   
2576         4           Wai Wai   
2577         3           Wai Wai   
2578         2           Wai Wai   
2579         1          Westbrae   

                                                Variety Style        Country  \
0                             T's Restaurant Tantanmen    Cup          Japan   
1     Noodles Spicy Hot Sesame Spicy Hot Sesame Guan...  Pack          China   
2                         Cup Noodles Chicken Vegetable   Cup  United States   
3                         GGE Ramen Snack Tomato Flavor  Pack          China   
4                                       Singapore Curry  Pack          India   
5                                Kimchi song Song Ramen  Pack    South Korea   
6                   Spice Deli Tantan Men With Cilantro   Cup          Japan   
7                                 Nabeyaki Kitsune Udon  Tray          Japan   
8                              Hokkaido Soy Sauce Ramen  Pack          Japan   
9                 The Original Spicy Stir-Fried Noodles  Pack      Singapore   
10                          Creamy tom Yum Kung Flavour  Pack       Thailand   
11                              Yokohama Tonkotsu Shoyu  Pack  United States   
12                   Mr. Bibim Stir-Fried Kimchi Flavor  Pack    South Korea   
13                         Deka Buto Kimchi Pork Flavor  Bowl          Japan   
14    Demae Ramen Bar Noodle Aka Tonkotsu Flavour In...  Pack          China   
15                     Mushroom Flavour Instant Noodles   Cup      Singapore   
16                                 Mario Masala Noodles  Pack          India   
17                                    Tokyo Shoyu Ramen  Pack  United States   
18                                           Mi Hai Cua  Pack        Vietnam   
19                                      Premium Gomtang  Pack    South Korea   
20                            Cup Noodles Laksa Flavour   Cup          China   
21           

### <span style="background-color: #F9F2EB">Your dataset description here: ... 
Review number is useless; a lot of varieties (unique so pretty useless may I say); non unique brands and styles; countries
Interesting things to write in this section include: 
- It is a dataset about ramen ratings with 2580 entries. 
- The dataset contains information about the brand, the country of origin, the varities, style, the star ratings and whether it has been part of the world top ten ramen. 
- The 'Review #' columnn is simply an index and has no relevance in the actual dataset. 
- 'Brand', 'Country', 'Variety' and 'Style' have categorical values. Specifically, 'Variety' have unique values. 
- The 'Star' rating is continuous decimal numbers out of 5, the majority (around 50%) of the ratings are above 3.0
- Over 50% of the 'Style' entries are 'Packs'
- The 'Top Ten' entries are NaN (null) if they have never been a Top Ten ramen. This field include the year of the reward, as well as the ranking. 

###### Take a look at the 'Top Ten' column. The row has a NaN value if the ramen has never been ranked. In datascience, NaN values are dangerous and can stop algorithms from working properly or working at all. First replace the NaN values by 'no'. 

In [17]:
# =========== YOUR CODE HERE ========
df['Top Ten']=df['Top Ten'].fillna('no',inplace=True)

In [18]:
df

Review #             Brand  \
0         2580         New Touch   
1         2579          Just Way   
2         2578            Nissin   
3         2577           Wei Lih   
4         2576    Ching's Secret   
5         2575     Samyang Foods   
6         2574           Acecook   
7         2573       Ikeda Shoku   
8         2572        Ripe'n'Dry   
9         2571              KOKA   
10        2570       Tao Kae Noi   
11        2569          Yamachan   
12        2568          Nongshim   
13        2567            Nissin   
14        2566            Nissin   
15        2565              KOKA   
16        2564              TRDP   
17        2563          Yamachan   
18        2562          Binh Tay   
19        2561             Paldo   
20        2560            Nissin   
21        2559  Jackpot Teriyaki   
22        2558              KOKA   
23        2557          Yamachan   
24        2556    Ching's Secret   
25        2555     Samyang Foods   
26        2554           Yum-Mie   
27        2553            Nissin   
28        2552           MyKuali   
29        2551              KOKA   
...        ...               ...   
2550        30             Myojo   
2551        29          Mee Jang   
2552        28            Nissin   
2553        27            Nissin   
2554        26            Nissin   
2555        25            Nissin   
2556        24            Nissin   
2557        23            Nissin   
2558        22          Nongshim   
2559        21          Nongshim   
2560        20          Nongshim   
2561        19          Nongshim   
2562        18            Ottogi   
2563        17         Quickchow   
2564        16           Samyang   
2565        15           Samyang   
2566        14           Samyang   
2567        13   Sapporo Ichiban   
2568        12   Sapporo Ichiban   
2569        11       Six Fortune   
2570        10             Smack   
2571         9             Sutah   
2572         8            Tung-I   
2573         7           Ve Wong   
2574         6             Vifon   
2575         5             Vifon   
2576         4           Wai Wai   
2577         3           Wai Wai   
2578         2           Wai Wai   
2579         1          Westbrae   

                                                Variety Style        Country  \
0                             T's Restaurant Tantanmen    Cup          Japan   
1     Noodles Spicy Hot Sesame Spicy Hot Sesame Guan...  Pack          China   
2                         Cup Noodles Chicken Vegetable   Cup  United States   
3                         GGE Ramen Snack Tomato Flavor  Pack          China   
4                                       Singapore Curry  Pack          India   
5                                Kimchi song Song Ramen  Pack    South Korea   
6                   Spice Deli Tantan Men With Cilantro   Cup          Japan   
7                                 Nabeyaki Kitsune Udon  Tray          Japan   
8                              Hokkaido Soy Sauce Ramen  Pack          Japan   
9                 The Original Spicy Stir-Fried Noodles  Pack      Singapore   
10                          Creamy tom Yum Kung Flavour  Pack       Thailand   
11                              Yokohama Tonkotsu Shoyu  Pack  United States   
12                   Mr. Bibim Stir-Fried Kimchi Flavor  Pack    South Korea   
13                         Deka Buto Kimchi Pork Flavor  Bowl          Japan   
14    Demae Ramen Bar Noodle Aka Tonkotsu Flavour In...  Pack          China   
15                     Mushroom Flavour Instant Noodles   Cup      Singapore   
16                                 Mario Masala Noodles  Pack          India   
17                                    Tokyo Shoyu Ramen  Pack  United States   
18                                           Mi Hai Cua  Pack        Vietnam   
19                                      Premium Gomtang  Pack    South Korea   
20                            Cup Noodles Laksa Flavour   Cup          China   
21           

###### The dataset includes the year and the ranking of the ramen (if ever ranked). However, it is done in a very unorganized manner (in my opinion). Therefore, replace all the entries by removing the ranking, and keeping ONLY the year. 
eg. "2016 #10" becomes "2016"

In [ ]:
# =========== YOUR CODE HERE ========


###### Uh oh, it seems like there are still some invalid values in the Top Ten column. Replace the '\n' values with 'no' as well. 

In [ ]:
# =========== YOUR CODE HERE ========

### Now, let's look at the following scenarios and help out Bobby in his journey. 

**1)** Bobby is a new food critique, and he wants to impress his boss with his favourite food: ramen. Help Bobby rank all of the countries based on their average ramen ratings, so he knows where to visit for his first professional critique. Plot it out in an ascending order. 
Explain your choice of the type of graph. 

In [ ]:
# =========== YOUR CODE HERE ======== 

### <span style="background-color: #F9F2EB"> Choice of graph type: ... </span>

**2)** Now that you have figured out countries' rankings, which country should Bobby visit to get the BEST ramens? 

In [ ]:
# =========== YOUR CODE HERE ========

**3)** Alright alright alright. There are over 2500 entries in this dataset. Bobby's old cellphone keeps lagging, every time he loads it. Help him extract the relevant data entries (the one from the best country) so he can navigate better. Oh while you're at it, rank the ramen in this country in ascending order according to their rankings! <br>
[hint: never overwrite your original dataset because you never know when you'll need it again.]

In [ ]:
# =========== YOUR CODE HERE ========

**4)** As personal preference, Bobby likes 'Pack' ramen. Therefore, he decides to start off with highest rated Pack ramen. Which one would it be? 

In [ ]:
# =========== YOUR CODE HERE ========

**5)** Bobby finishes his critique and shows it to his boss. Unfortunately, his boss was not impressed! They said that the brand was too mainstream, and too many critiques have already wrote articles about it. They told Bobby to target the ramen with a 5 star rating that is from either Japan, Malaysia, Singapore or South Korea. AND that have ranked 'Top Ten'. What are the possible ramens? 

Can you do this in one single line? 

In [ ]:
# =========== YOUR CODE HERE ========

**6)** Bobby starts his ramen eating journey, and arrives to Singapore. In a single day finishes ALL the ramens above 3.0 (inclusive) stars, that are NOT in a 'Pack'. (Yes, he eats a lot). Change all of the 'Tries' entries of these rows to True. 

In [ ]:
# =========== YOUR CODE HERE ========

**7)** Bobby goes into the Singaporian 7-eleven and stumbles upon a new ramen that he has never heard of, and was not part of the dataset. He tried it and absolutely LOVED it (5/5). It was the packed 'Beef Noodle Soup' flavoured ramen from 'Prima Taste'. Help him add it into the dataset! 

In [ ]:
# =========== YOUR CODE HERE ========

**8)** Once again, he notices that his phone is super slow. Therefore, he decides to delete each ramen he has tried from the list, instead of marking 'Tried' as True. This extra column makes the phone even slower. Help remove the 'Tried' column, and remove the colums where 'Tries' is True. Try to make this inplace so that. This will help Bobby save RAM, as it will not create a temporary duplicate dataset on his phone! 

In [ ]:
# =========== YOUR CODE HERE ========

**9)** It suddenly hit Bobby. Why are there so many different styles of ramen? He is curious of the distribution of the packaging styles of the remaining to try ramens. Help him visualize the percentage of occurrence of each package. Explain the choice of your graph style. 

In [ ]:
# =========== YOUR CODE HERE ========

### <span style="background-color: #F9F2EB"> Choice of graph type ... </span> 

**10)** After trying all the Singaporian ramen, he decides that it is time for him to change to a more powerful phone. He wants to save his UPDATED dataset to an external SD card so he can transfer it to his new phone. Save the dataset as 'Name-Surname-ramen-to-try.csv'. 

Don't forget to update the dataset first! Make sure that you save don't also save the row indeces! 

In [ ]:
# =========== YOUR CODE HERE ======== 

### Data Visualization 

###### 1) Load ./dataset/shiba-inu.jpg and display it.

In [ ]:
# =========== YOUR CODE HERE ======== 

###### 2) Display the shiba-inu in (100, 120) size.

In [ ]:
# =========== YOUR CODE HERE ======== 

###### 3) Load 'star-boo.jpg', 'baby-dog.jpg' and 'chihuahua.jpg' in the dataset folder and display all four images in a 2x2 grid. 

In [ ]:
# =========== YOUR CODE HERE ======== 

###### 4) Give the image a title of your choice. Remove the axis from the pictures and redisplay the pictures of the cute dogs. 

In [ ]:
# =========== YOUR CODE HERE ======== 

## HAVE FUN!!